From: Christian Hergert <chergert@redhat.com>
Date: Wed, 28 Aug 2024 17:02:07 -0700
Subject: gnome-bg: only parse XML files as slideshows

Currently, loading gnome-control-center tries to load every single file
in your wallpaper collection about 10x into memory, then parse it with
GMarkup only to discard it when it fails.

This culls that to 2 loads of actual XML files on my local jhbuild setup.

Bug: https://gitlab.gnome.org/GNOME/gnome-control-center/-/issues/3167
Origin: upstream, 42, commit:f27670d498b47695bf76b7ff8d2df067c87b3038
---
 libgnome-desktop/gnome-bg/gnome-bg.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/libgnome-desktop/gnome-bg/gnome-bg.c b/libgnome-desktop/gnome-bg/gnome-bg.c
index a95d437..bac40de 100644
--- a/libgnome-desktop/gnome-bg/gnome-bg.c
+++ b/libgnome-desktop/gnome-bg/gnome-bg.c
@@ -2218,6 +2218,18 @@ read_slideshow_file (const char *filename,
 {
         GnomeBGSlideShow *show;
 
+        /* Ignore anything not an XML file because we don't want to waste
+         * time reading that file into memory only to try to parse it with
+         * GMarkup.
+         */
+        if (!g_str_has_suffix (filename, ".xml")) {
+                g_set_error_literal (err,
+                                     G_IO_ERROR,
+                                     G_IO_ERROR_NOT_SUPPORTED,
+                                     "Not a slideshow file");
+                return NULL;
+        }
+
         show = gnome_bg_slide_show_new (filename);
 
         if (!gnome_bg_slide_show_load (show, err)) {
